bots/topological/src/main/java/org/openjdk/skara/bots/topological/TopologicalBot.java

```diff
@@ -0,0 +1,220 @@
+                        var parent = parentsIt.next();
+                        try {
+                            mergeIfAhead(repo, branch, parent);
+                            progress = true;
+                            parentsIt.remove(); // avoid doing pointless merges
+                        } catch(IOException e) {
+                            log.severe("Merge with " + parent + " failed. Reverting...");
+                            repo.abortMerge();
+                            failedMerges.add(branch + " <- " + parent);
+                            failed = true;
+                        }
+                    }
+                } while(progress && failed);
+
+                if (!failedMerges.isEmpty()) {
+                    throw new IOException("There were failed merges:\n" + failedMerges);
+                }
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+        log.info("Ending topobot run");
+    }
+
+    private static Stream<Branch> dependencies(Path depsFile) throws IOException {
+        if (Files.exists(depsFile)) {
+            var lines = Files.readAllLines(depsFile).stream().filter(s -> !s.isEmpty()).collect(Collectors.toList());
+            if (lines.size() > 1) {
+                throw new IllegalStateException("Multiple non-empty lines in " + depsFile.toString() + ": "
+                        + String.join("\n", lines));
+            }
+            return Stream.of(lines.get(0).split(" ")).map(Branch::new);
+        } else {
+            return Stream.of(new Branch("master"));
+        }
+    }
+
+    private List<Branch> orderedBranches(Repository repo, Path depsFile) throws IOException {
+        List<Edge> deps = new ArrayList<>();
+        for (var branch : branches) {
+            repo.checkout(branch);
+            dependencies(depsFile).forEach(dep -> deps.add(new Edge(dep, branch)));
+        }
+        return TopologicalSort.tsort(deps).stream()
+            .filter(branch -> !branch.name().equals("master"))
+            .collect(Collectors.toList());
+    }
+
+    private void mergeIfAhead(Repository repo, Branch branch, Branch parent) throws IOException {
+        var fromHash = repo.resolve(parent.name()).orElseThrow();
+        var oldHead = repo.head();
+        if (!repo.contains(branch, fromHash)) {
+            var isFastForward = repo.isAncestor(oldHead, fromHash);
+            repo.merge(fromHash);
+            if (!isFastForward) {
+                log.info("Merged " + parent + " into " + branch);
+                repo.commit("Automatic merge with " + parent, "duke", "duke@openjdk.org");
+            } else {
+                log.info("Fast forwarded " + branch + " to " + parent);
+            }
+            log.info("merge with " + parent + " succeeded. The following commits will be pushed:\n"
+                    + log(repo, "origin/" + branch.name(), branch.name()).collect(Collectors.joining("\n", "\n", "\n")));
```

Comment:
You could just do `repo.commits("origin/" + branch.name() + ".." + branch.name())` instead of rolling your own